### DISCUSSION TITLE
Serving a release WASM app causes rendering to fail

### DISCUSSION THREAD

**Discussion by mystborn** [AvaloniaUI/Avalonia#13378] [Q&A]
## Describe the bug
When attempting to serve a release build of the project that I'm working on, the rendering fails, only showing the class name of a view in the visual tree. Serving the debug folder instead yields the expected results (using `dotnet serve` in both cases for consistency).

## To Reproduce

```sh
dotnet workload restore
dotnet publish -c Release
dotnet serve --directory .\bin\Release\net7.0\browser-wasm\AppBundle\
```

## Screenshots
Debug Mode
![image](https://github.com/AvaloniaUI/Avalonia/assets/10185859/81419e80-9473-4904-bcc2-7fc8abb9cf06)

Release Mode
![image](https://github.com/AvaloniaUI/Avalonia/assets/10185859/0df691af-c556-468e-9449-0ddfa9f3fcec)

## Environment

- OS: Browser
- Avalonia-Version: 11.0.2

**Comment by maxkatz6** [comment]
You found that dotnet Browser is built with trimming always enabled in the release configuration.
And your views were not rooted by the compiler (probably used only via reflection). Could it be view locator?

**Reply by maxkatz6** [reply]
cc @timunie had some examples without reflection based view locator IIRC

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/DataTemplates/IDataTemplateSample 
or
https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/DataTemplates/FuncDataTemplateSample

**Reply by mystborn** [reply]
Is there a way to disable trimming?

**Reply by maxkatz6** [reply]
On browser only in debug mode IIRC

**Reply by maxkatz6** [reply]
In general, see https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-7-0

**Comment by dietelTiMaMi** [comment]
Hello I'm running into similar Problems. It seems my whole views and viewModels were trimmed away.
is there a way to root a whole assembly to prevent it from being trimmed away?

**Reply by timunie** [reply]
See comments and links above. In general, this is a question about .NET and not really Avalonia specific.

**Comment by AlexStanglEmerson** [comment]
Here's the method I use to fix this issue since I didn't see it mentioned on Microsoft's documentation page. Annotate your `Main()` entry point with `DynamicDependency` for the classes that are getting trimmed away.

```csharp
[assembly: SupportedOSPlatform("browser")]

internal sealed partial class Program
{
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(Tab1View))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(Tab2View))]
    private static async Task Main(string[] args)
    {
        ...
    }

    ...
}
```
